src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -1375,64 +1482,207 @@ krb5_read_tkt_policy(krb5_context context, krb5_ldap_context *ldap_context,
     return st;
 }
 
-krb5_error_code
-krb5_decode_krbsecretkey(krb5_context context, krb5_db_entry *entries,
-                         struct berval **bvalues,
-                         krb5_tl_data *userinfo_tl_data, krb5_kvno *mkvno)
+struct key_entry {
+    krb5_int16 n_key_data;
+    krb5_key_data *key_data;
+};
+
+/*
+ * Decode keys from ldap search results.
+ *
+ * Arguments:
+ *  - bvalues
+ *      The ldap search results containing the key data.
+ *  - mkvno
+ *      The master kvno that the keys were encrypted with.
+ *  - keys_out
+ *      The decoded keys in a key_entry struct. Must be freed.
+ *  - n_keys_out
+ *      The number of key_entry's in keys_out.
+ *  - total_keys_out
+ *      An optional argument that if given will be set to the total number of
+ *      keys found throughout all the entries: sum(keys_out.n_key_data)
+ *      May be NULL.
+ */
+static krb5_error_code
+decode_keys(struct berval **bvalues, krb5_kvno *mkvno,
+            struct key_entry **keys_out, unsigned int *n_keys_out,
+            /* NULL */ unsigned int *total_keys_out)
 {
-    char                        *user=NULL;
-    int                         i=0, j=0, noofkeys=0;
-    krb5_key_data               *key_data=NULL, *tmp;
-    krb5_error_code             st=0;
+    krb5_error_code err = 0;
+    unsigned int n_keys, i, ki, total_keys;
+    struct key_entry *p_keys = NULL;
+
+    *keys_out = NULL;
+    *n_keys_out = 0;
+    if (total_keys_out)
+        *total_keys_out = 0;
 
-    if ((st=krb5_unparse_name(context, entries->princ, &user)) != 0)
+    /*
+     * Get the number of keys. Previously, similar implementations used realloc
+     * instead of pre-counting but I think this will have better performance.
+     */
+    for (n_keys = 0, i = 0; bvalues[i] != NULL; i++)
+        if (bvalues[i]->bv_len > 0)
+            n_keys++;
+
+    p_keys = k5calloc(n_keys, sizeof(struct key_entry), &err);
+    if (p_keys == NULL)
         goto cleanup;
 
-    for (i=0; bvalues[i] != NULL; ++i) {
-        krb5_int16 n_kd;
-        krb5_key_data *kd;
+    for (i = 0, ki = 0, total_keys = 0; bvalues[i] != NULL; i++) {
         krb5_data in;
 
         if (bvalues[i]->bv_len == 0)
             continue;
+
         in.length = bvalues[i]->bv_len;
         in.data = bvalues[i]->bv_val;
 
-        st = asn1_decode_sequence_of_keys (&in,
-                                           &kd,
-                                           &n_kd,
+        err = asn1_decode_sequence_of_keys(&in,
```

Comment:
Minimize the function argument overwrapping here; I believe this call can fit on two lines.  (The old code was just as much at fault, but it should be fixed while we're changing it.)